cpp.mk

```diff
@@ -79,11 +79,26 @@ CLEAN_DIRS += $${$1_OBJECT_DIR}
 clangtidy_$1: $${$1_CLANGTIDY}
 	cat $${$1_CLANGTIDY}
 
+# Service tests must run under Docker, to provide isolation, and this Makefile
+# enforces this. To run service tests, repositories must provide:
+# - a service_tests/ directory at the top level
+# - a Dockerfile in that directory which will run the tests
+# - a copy_files_to_docker_context.sh script in that directory, which is
+#   responsible for copying any necessary files (e.g. the binary and libraries
+#   under test) into the service_test/ directory structure - the Dockerfile
+#   cannot see anything above this directory.
 service_test:
 ifeq ($2,release)
 .PHONY: service_test_$1
 service_test_$1: $${BUILD_DIR}/bin/$1
-	if [ -d $${SERVICE_TEST_DIR} ]; then $${SERVICE_TEST_DIR}/service_test_main; else echo "No service tests found"; fi
+	if [ -d $${SERVICE_TEST_DIR} ]; then \
+		cd $${SERVICE_TEST_DIR} && \
+		bash -xe ./copy_files_to_docker_context.sh && \
+		docker build -t $(shell whoami)-$$@ . && \
+		docker run -t $(shell whoami)-$$@; \
```

Comment:
Why do we need to allocate a TTY (which is what `-t` does - in `docker build` it creates a tag, in `docker run` it creates a psuedo-TTY for the container)?

We should probably be passing `--rm` otherwise we are going to leave containers lying around.

We should probably also considering doing `docker rmi $(cat $${BUILD_DIR}/service_test_docker_image)` to delete the Docker image afterwards, although that will break the caching of the Docker image.